sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala

```diff
@@ -497,4 +587,121 @@ class TransformWithStateInitialStateSuite extends StateStoreMetricsTest
+    val df_joined =
+      valueDf.unionByName(unflattenListDf, true)
+        .unionByName(unflattenMapDf, true)
+    val kvDataSet = inputData.toDS().groupByKey(x => x.key)
+    val initDf = df_joined.as[UnionUnflattenInitialStateRow].groupByKey(x => x.groupingKey)
+    kvDataSet.transformWithState(
+      new InitialStatefulProcessorWithUnflattenStateDataSource(),
+      TimeMode.None(), OutputMode.Append(), initDf).toDF()
+  }
+
+  private def testInitialStateWithStateDataSource(
+      flattenOption: Boolean)
+      (startQuery: (DataFrame, DataFrame, DataFrame,
+        MemoryStream[InitInputRow]) => DataFrame): Unit = {
+    Seq(("5", "2"), ("5", "8"), ("5", "5")).foreach { partitions =>
+      test("transformWithStateWithInitialState - state data source reader dataframe " +
+        s"as initial state with flatten option set to $flattenOption, the first stream and " +
+        s"the second stream is running on shuffle partition number of ${partitions._1} and " +
+        s"${partitions._2} respectively.") {
+        withSQLConf(SQLConf.STATE_STORE_PROVIDER_CLASS.key ->
+          classOf[RocksDBStateStoreProvider].getName) {
+          withTempDir { checkpointDir =>
+            SQLConf.get.setConfString(SQLConf.SHUFFLE_PARTITIONS.key, partitions._1)
+            val inputData = MemoryStream[String]
+            val result = inputData.toDS()
+              .groupByKey(x => x)
+              .transformWithState(new StatefulProcessorWithAllStateVars(),
+                TimeMode.None(),
+                OutputMode.Update())
+
+            testStream(result, OutputMode.Update())(
+              StartStream(checkpointLocation = checkpointDir.getCanonicalPath),
+              AddData(inputData, "a", "b"),
+              CheckNewAnswer(("a", "1"), ("b", "1")),
+              AddData(inputData, "a", "b", "a"),
+              CheckNewAnswer(("a", "3"), ("b", "2"))
+            )
+
+            // state data source reader for state vars
+            val valueDf = spark.read
+              .format("statestore")
+              .option(StateSourceOptions.PATH, checkpointDir.getAbsolutePath)
+              .option(StateSourceOptions.STATE_VAR_NAME, "countState")
+              .load()
+
+            val listDf = spark.read
+              .format("statestore")
+              .option(StateSourceOptions.PATH, checkpointDir.getAbsolutePath)
+              .option(StateSourceOptions.STATE_VAR_NAME, "listState")
+              .option(StateSourceOptions.FLATTEN_COLLECTION_TYPES, flattenOption)
+              .load()
+
+            val mapDf = spark.read
+              .format("statestore")
+              .option(StateSourceOptions.PATH, checkpointDir.getAbsolutePath)
+              .option(StateSourceOptions.STATE_VAR_NAME, "mapState")
+              .option(StateSourceOptions.FLATTEN_COLLECTION_TYPES, flattenOption)
+              .load()
+
+            // create a df where each row contains all value, list, map state rows
```

Comment:
Honestly I feel like the test is non-trivially complicated - someone has to understand multiple tws implementations and connect them how values will transform.

That's unnecessary for verifying the behavior - we just need to have the state in the checkpoint (first query), and load that state as initial state in another query. No need to complicate building the state in the first query at all.

For example, in StatefulProcessorWithAllStateVars, just handle actions against state variables so that we intuitively know about current state variables with inputs, rather than tracking how values are set in value state / list state / map state.

Reply:
Not sure if I understand your comments - are you suggesting we should write a simpler implementation of `StatefulProcessorWithAllStateVars` for testing purpose? The current implementation is a counter where we keep the count of occurrence of input rows in the value, list, map state respectively, not sure if we can make this even simpler to update three types of state variables in a single stateful processor. I added the explanation of what the stateful processor is doing in the class docstring and I also added a few lines of comments in the test case to explain why we are doing this complex test case - it is mainly to mimic user case on how they will potentially use the data source integration with tws. I added some comments in `startQueryWithDataSourceDataframeAsInitState` explaining the difference of flattenOption as well. Let me know if this makes sense to you.

Reply:
If someone has to verify the test to work correctly, they have to do the following

1) learn how StatefulProcessorWithAllStateVars works
2) go through the input rows and calculate by themselves how the values of three state vars would look like
3) look at how the second query is constructed with initial state from state data source reader
4) learn how InitialStatefulProcessorWithStateDataSource works
5) reason about the value of state variables for initial state
6) go through the input rows and calculate by themselves how the values of three state vars would look like

I understand the purpose, making the test to be more e2e usage like, but this is a unit test, and what we describe for test name is about `ability of setting up initial state with the result of state data source reader`. I don't expect the test to show up the flexibility of what users can do. That should be placed to the examples directory or some blog post with github repo, not here.

I think the test still verifies what it intends to do, if the test just shows the ability of migration for checkpoint "with the same logic of the query". The test doesn't need for reader to understand two different queries; if the test runs for a couple batches, and can continue with another checkpoint (meaning batch 0) with retaining state via initial state (and allow further modification), that should be enough.

Reply:
Technically speaking, we should already have multiple unit tests partially testing the part for each and chaining them should test exactly the same. Otherwise we are just missing some unit tests in prior.

Reply:
I just modified the stateful processor for the second stateful processor to the simplest put of count. I guess one reason that the suite is complicated is that we need to have coverage for value, list, map state in a single processor to have the coverage on all composite types. Do you think we should only keep it simple by only testing on one of the state variable?

Reply:
One thing I want to remind is, we can just verify the value of state variable, with state data source reader. We no longer need to have weird types of inputs to trigger retrieval of the value for the state variable. No longer have to have the second stateful processor. Just query them with state data source reader, from `Execute` command.

That said, you can retain the first stateful processor, with adding ability to handle initial state. After that, I don't see any technical challenge to verify the behavior with a single query from code perspective. (Technically they should be two different unique streaming queries but should be able to use the same code.)

I know there are existing tests having such a complicated stateful processor - we can simplify them as of now. (Definitely not saying this PR has to do that.)

Reply:
Removed one of the stateful processor and check the initial state handling in `Exceute` block.